azure/services/subnets/spec.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subnets
+
+import (
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-02-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+)
+
+// SubnetSpec defines the specification for a Subnet.
+type SubnetSpec struct {
+	Name              string
+	ResourceGroup     string
+	SubscriptionID    string
+	CIDRs             []string
+	VNetName          string
+	IsVNetManaged     bool
+	RouteTableName    string
+	SecurityGroupName string
+	Role              infrav1.SubnetRole
+	NatGatewayName    string
+}
+
+// ResourceName returns the name of the subnet.
+func (s *SubnetSpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *SubnetSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName returns the name of the VNet that owns this subnet.
+func (s *SubnetSpec) OwnerResourceName() string {
+	return s.VNetName
+}
+
+// Parameters returns the parameters for the subnet.
+func (s *SubnetSpec) Parameters(existing interface{}) (parameters interface{}, err error) {
+	if existing != nil {
+		existingSubnet, ok := existing.(network.Subnet)
+		if !ok {
+			return nil, errors.Errorf("%T is not a network.Subnet", existing)
+		}
+
+		// If the subnet infrav1 spec has no CIDRs but the existing subnet has CIDRs, update the spec
+		if len(s.CIDRs) == 0 { // empty slice or nil?
+			// This is a bit of a workaround. By returning, we force a CreateOrUpdate call, which will trigger
+			// UpdateSubnetSpec() to be called. That will update the infrav1 spec, and those changes will be
+			// reflected in this spec when SubnetSpecs() is called.
+			return existingSubnet, nil
+		}
+
+		return nil, nil
+	} else if !s.IsVNetManaged {
+		// TODO: should this be a terminal error to break out of the reconcile loop?
```

Comment:
I think it seems reasonable to return a terminal error here. WDYT, @CecileRobertMichon.

Reply:
Sounds good, I opened #1973. I can open a separate PR to support terminal errors in the AzureCluster controller and add a terminal error here.